crates/matrix-sdk/src/room/reply.rs

```diff
@@ -0,0 +1,404 @@
+
+                    let mut content = if is_reply == ReplyWithinThread::Yes {
+                        content.make_reply_to_raw(
+                            &raw_event,
+                            replied_to_info.event_id.to_owned(),
+                            room_id,
+                            ForwardThread::No,
+                            mention_the_sender,
+                        )
+                    } else {
+                        content.into()
+                    };
+
+                    let thread_root = if let Some(EncryptedRelation::Thread(thread)) =
+                        previous_content.as_ref().and_then(|c| c.relates_to.as_ref())
+                    {
+                        thread.event_id.to_owned()
+                    } else {
+                        replied_to_info.event_id.to_owned()
+                    };
+
+                    let thread = if is_reply == ReplyWithinThread::Yes {
+                        Thread::reply(thread_root, replied_to_info.event_id)
+                    } else {
+                        Thread::plain(thread_root, replied_to_info.event_id)
+                    };
+
+                    content.relates_to = Some(Relation::Thread(thread));
+                    content
+                }
+
+                EnforceThread::MaybeThreaded => content.make_reply_to_raw(
+                    &raw_event,
+                    replied_to_info.event_id,
+                    room_id,
+                    ForwardThread::Yes,
+                    mention_the_sender,
+                ),
+
+                EnforceThread::Unthreaded => content.make_reply_to_raw(
+                    &raw_event,
+                    replied_to_info.event_id,
+                    room_id,
+                    ForwardThread::No,
+                    mention_the_sender,
+                ),
+            }
+        }
+    };
+
+    Ok(content.into())
+}
+
+async fn replied_to_info_from_event_id<S: EventSource>(
+    source: S,
+    event_id: &EventId,
+) -> Result<RepliedToInfo, ReplyError> {
+    let event = source.get_event(event_id).await?;
+
+    let raw_event = event.into_raw();
+    let event = raw_event.deserialize().map_err(|_| ReplyError::FailedToDeserializeEvent)?;
+
+    let reply_content = match &event {
+        AnySyncTimelineEvent::MessageLike(event) => {
+            if let AnySyncMessageLikeEvent::RoomMessage(SyncMessageLikeEvent::Original(
+                original_event,
+            )) = event
+            {
+                ReplyContent::Message(original_event.content.clone())
+            } else {
+                ReplyContent::Raw(raw_event)
+            }
+        }
+        AnySyncTimelineEvent::State(_) => return Err(ReplyError::StateEvent),
+    };
+
+    Ok(RepliedToInfo {
+        event_id: event_id.to_owned(),
+        sender: event.sender().to_owned(),
+        timestamp: event.origin_server_ts(),
+        content: reply_content,
+    })
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I wasn't sure how many tests to add here since we already have the tests in `replies.rs`.

Reply:
It would be sweet to have a smoke test + have tests that reach each `ReplyError` variant. (We can also *remove* some tests from the timeline's code, if they were only testing the code moved here; otherwise the timeline is now testing this layer, and that's a layer mismatch.)